NASA-IMPACT/veda-ui
# add initial playwright suite

This adds a playwright test suite that will work for the veda-ui pages generated with the mock data as well as allow the veda-config repo to use those same tests on all of the stories and datasets in the config repo. It does this by generating a json containing the names of all datasets and stories by parsing the `mdx` files.   All tests are written in a way that they can be run within the `veda-ui` repo leveraging the mdx files in the mocks directory. A separate PR in the `veda-config` will be able to run the same tests against the actual stories and datasets. https://github.com/NASA-IMPACT/veda-config/pull/363

Each tests ensures that the pages render with the name and that there are no javascript errors on those pages. 
The tests verify that all /catalog and /dataset routes load with tiles for the expected.
There is a test of the explore route for each dataset that ensures each one loads without a javascript error.
There is also a test flow that  makes a general search on the /analysis page.

Interactions with the pages are accomplished with locators defined as Page Objects in the `e2e/pages` directory.

A github workflow is included to run the tests against the mocks on all pull requests.  The test run will have a report with trace to allow for debugging of failed tests. That report is retained for 30 days.  **To successfully complete all tests, a github secret should be added with `MAPBOX_TOKEN`.  Without this, some pages attempting to render the mapbox maps will cause tests to fail.** 

Assumptions: in veda-ui, the mock data lives in the `/mocks` directory and all mdx files that are rendered into stories or catalogs have a `name` property and all datasets also have an `id`.


How to test
Run the yarn install to pull in the new dependencies. Run the script yarn test:e2e to verify that the tests run locally.** To verify that tests catch errors, modify pages to break them and ensure that the test suite finds the failure.

** you may be asked to install the playwright browsers on the first run of the test suite.
